**Pontia Tech - Master Data Analytics - Visualización de Datos con Python**

***
Datos del alumno (Nombre y Apellidos):María Lucía Fernández Morant

Fecha:14/01/25
***

<span style="font-size: 30pt; font-weight: bold; color: #0098cd;">Estudio Politica Descuentos Pontia SuperStore</span>

<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">Sesión Live Práctica</span>

Entregable Visualización Python

**Objetivos**

El objetivo de este caso práctico es realizar un estudio de como ha impactado la politica de descuentos a la rentabilidad de la empresa Pontia SuperStore. Para ello usaremos la librería Plotly para la elaboración de gráficos Interactivos.

**Enunciado**

Somos parte del equipo de datos de la empresa PontiaSuperStore han encargado hacer un estudio de cómo ha impactado la politica de descuentos en la rentabilidad de la empresa.

Para ello contamos con un dataset con los pedidos efectuados a la empresa. En cada pedido hay una columna que nos informa del % de descuento que lleva esa venta asociada. ¿Esta afectando a la rentabilidad de la empresa? ¿Es igual para todos los productos?

Elabora un reporte usando python que contenga diferentes visualizaciones ilustrando el problema.

Trata de no pensar directamente en la solución y construye una historia a traves de los datos.

**Requisitos de la actividad:**

1º El reporte debe contener mínimo tres visualizaciones. Recomendación: Es interesante pensar en una estructura introducción - nudo - desenlace. En la que se presenta el problema primero, se aportan diferentes puntos de vista y se da una solución.

2º Utilización de Plotly para resolver el ejercicio.

3º Entregar el Notebook con las respuestas argumentadas, no solo el gráfico.


<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">1. Importación Librerías y Carga Datos</span>

<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">1.1. Importación Librerías</span>


In [1]:
# Librerías Procesamiento de Datos.
import pandas as pd
import numpy as np

# Librería de Visualización
import plotly.express as px

# Avoid Warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
%matplotlib inline

<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">1.2. Carga de Datos</span>

La carga de los datos la podemos hacer en Colab usando la librería "io" o en local.

Google Colab

In [3]:
## En Google Colab
from google.colab import files
uploaded = files.upload()

Saving PontiaSuperStore (2).xls to PontiaSuperStore (2).xls


In [4]:
import io
# Get the actual filename from the 'uploaded' dictionary
actual_filename = list(uploaded.keys())[0]
# Read the Excel file using the actual filename
df = pd.read_excel(io.BytesIO(uploaded[actual_filename]))
df.head(3)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16  Second Class    DV-13045   

     Customer Name    Segment        Country         City  ... Postal Code  \
0      Claire Gute   Consumer  United States    Henderson  ...       42420   
1      Claire Gute   Consumer  United States    Henderson  ...       42420   
2  Darrin Van Huff  Corporate  United States  Los Angeles  ...       90036   

   Region       Product ID         Category Sub-Category  \
0   South  FUR-BO-10001798        Furniture    Bookcases   
1   South  FUR-CH-10000454        Furniture       Chairs   
2    West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.96         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.62         2   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  
2       0.0    6.8714  

[3 rows x 21 columns]

En local

In [ ]:
# Solamente en caso de no usar Google Colab.
# df = pd.read_excel("PontiaSuperStore.xls", sheet_name="Orders")

<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">1.3. Generación de variables calculadas.</span>

Genere a partir del dataframe cargado las siguientes variables calculadas.

- Sales/Quantity: Precio de Venta por unidad de producto en cada operación de venta
- Profit/Quantity: Beneficio por unidad de producto en cada operación de venta
- flag_descuento: Marca con "Con Descuento" si esa operacion de venta ha tenido descuento (esto es descuento>1), "Sin Descuento" si no.
- Year: Variable con el año de la fecha de la operación de venta.

Respuesta:

In [5]:
# 1. Sales/Quantity: Precio de Venta por unidad de producto en cada operación de venta
df['Sales quantity'] = df['Sales'] / df['Quantity']


In [6]:
df.head()


Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ... Region  \
0      Claire Gute   Consumer  United States        Henderson  ...  South   
1      Claire Gute   Consumer  United States        Henderson  ...  South   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   West   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

   Discount    Profit  Sales quantity  
0      0.00   41.9136        130.9800  
1      0.00  219.5820        243.9800  
2      0.00    6.8714          7.3100  
3      0.45 -383.0310        191.5155  
4      0.20    2.5164         11.1840  

[5 rows x 22 columns]

In [7]:
# 2. Profit/Quantity: Beneficio por unidad de producto en cada operación de venta
df['unit profit'] = df['Profit'] / df['Quantity']

In [8]:
df.head ()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit  Sales quantity  unit profit  
0     0.00   41.9136        130.9800      20.9568  
1     0.00  219.5820        243.9800      73.1940  
2     0.00    6.8714          7.3100       3.4357  
3     0.45 -383.0310        191.5155     -76.6062  
4     0.20    2.5164         11.1840       1.2582  

[5 rows x 23 columns]

In [9]:
# 3. flag_descuento: Marca con "Con Descuento" si esa operacion de venta ha tenido descuento (esto es descuento>1), "Sin Descuento" si no.

df['flag_descuento'] = df['Discount'].apply(lambda x: 'Con Descuento' if x > 0 else 'Sin Descuento')


In [10]:
df.head ()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

          Category  Sub-Category  \
0        Furniture     Bookcases   
1        Furniture        Chairs   
2  Office Supplies        Labels   
3        Furniture        Tables   
4  Office Supplies       Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit  Sales quantity  unit profit  flag_descuento  
0     0.00   41.9136        130.9800      20.9568   Sin Descuento  
1     0.00  219.5820        243.9800      73.1940   Sin Descuento  
2     0.00    6.8714          7.3100       3.4357   Sin Descuento  
3     0.45 -383.0310        191.5155     -76.6062   Con Descuento  
4     0.20    2.5164         11.1840       1.2582   Con Descuento  

[5 rows x 24 columns]

In [11]:
# 4. Año de la fecha (Year)
df['Year'] = pd.to_datetime(df['Order Date']).dt.year

df.head ()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Sub-Category                                       Product Name     Sales  \
0    Bookcases                  Bush Somerset Collection Bookcase  261.9600   
1       Chairs  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400   
2       Labels  Self-Adhesive Address Labels for Typewriters b...   14.6200   
3       Tables      Bretford CR4500 Series Slim Rectangular Table  957.5775   
4      Storage                     Eldon Fold 'N Roll Cart System   22.3680   

  Quantity Discount    Profit Sales quantity  unit profit  flag_descuento  \
0        2     0.00   41.9136       130.9800      20.9568   Sin Descuento   
1        3     0.00  219.5820       243.9800      73.1940   Sin Descuento   
2        2     0.00    6.8714         7.3100       3.4357   Sin Descuento   
3        5     0.45 -383.0310       191.5155     -76.6062   Con Descuento   
4        2     0.20    2.5164        11.1840       1.2582   Con Descuento   

   Year  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015  

[5 rows x 25 columns]

In [12]:
df.info ()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          9994 non-null   int64         
 1   Order ID        9994 non-null   object        
 2   Order Date      9994 non-null   datetime64[ns]
 3   Ship Date       9994 non-null   datetime64[ns]
 4   Ship Mode       9994 non-null   object        
 5   Customer ID     9994 non-null   object        
 6   Customer Name   9994 non-null   object        
 7   Segment         9994 non-null   object        
 8   Country         9994 non-null   object        
 9   City            9994 non-null   object        
 10  State           9994 non-null   object        
 11  Postal Code     9994 non-null   int64         
 12  Region          9994 non-null   object        
 13  Product ID      9994 non-null   object        
 14  Category        9994 non-null   object        
 15  Sub-

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">2. Estudio Descuentos PontiaSuperStore</span>

Elaboración del reporte.

Argumenta y explica la motivación del gráfico así como sus enunciados.

<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">2.1. Gráfico 1.</span>

Respuesta:

In [13]:
#comparativa de ventas con y sin descuento

In [14]:
con_sin_desc = df.groupby('flag_descuento').agg(
    Count=('Sales', 'count')
).reset_index()
con_sin_desc

flag_descuento  Count
0  Con Descuento   5196
1  Sin Descuento   4798

In [15]:
import plotly.express as px

fig = px.bar(
    con_sin_desc,
    x='flag_descuento',
    y='Count',
    title='Cantidad de Ventas según el Descuento',
    labels={'flag_descuento': 'Categoría de Descuento', 'Count': 'Cantidad de Ventas'},
    color='flag_descuento',
    template = 'simple_white',
)
fig.show()

En este gráfico se puede apreciar que la cantidad de ventas con descuentos es de 5196 y sin descuentos es de 4798, con lo cual vemos el el porcentage no es muy significativo.


In [16]:
# La misma de antes pero separando por años

In [17]:
con_años = df.groupby(['Year', 'flag_descuento']).agg(
    Count=('Sales', 'count')  # Cantidad de ventas
).reset_index()

con_años

Year flag_descuento  Count
0  2014  Con Descuento   1056
1  2014  Sin Descuento    937
2  2015  Con Descuento   1079
3  2015  Sin Descuento   1023
4  2016  Con Descuento   1339
5  2016  Sin Descuento   1248
6  2017  Con Descuento   1722
7  2017  Sin Descuento   1590

In [18]:
fig1 = px.bar(
    con_años,
    x='Year',
    y='Count',
    color='flag_descuento',
    barmode='group',
    title='Cantidad de Ventas según Descuento y Año',
    labels={'flag_descuento': 'Categoría de Descuento', 'Count': 'Cantidad de Ventas', 'Year': 'Año'},
    color_discrete_map={'Con Descuento': 'skyblue', 'Sin Descuento': 'orange'},
    template = 'simple_white',
)

fig1.show()

En este gráfico mostramos las ventas con y sin descuento por años.
En el comprobamos que las ventas con descuento es mayor que las ventas sin descuento.
Comprobamos que las ventas han ido en aumento en ambas,
pero se ve  que la empresa tiene muchos productos en oferta y pueda ser el motivo de que las ventas con descuento sean mayores.

Conclusiones: La empresa ha seguido la tendencia de tener una gran parte de su catalogo con descuento aplicado



<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">2.2. Gráfico 2.</span>

Respuesta:

In [19]:
comparativa = df.groupby(['Year', 'flag_descuento']).agg(
    Total_Sales=('Sales', 'sum'),  # Sumar las ventas por año y descuento
    Total_Profit=('Profit', 'sum')  # Sumar los beneficios por año y descuento
).reset_index()

comparativa

Year flag_descuento  Total_Sales  Total_Profit
0  2014  Con Descuento  269287.7681    -9073.1924
1  2014  Sin Descuento  214959.7300    58617.1665
2  2015  Con Descuento  244012.6590    -7252.1529
3  2015  Sin Descuento  226519.8500    68870.7566
4  2016  Con Descuento  311594.8580    -7546.7936
5  2016  Sin Descuento  297610.7400    89341.9679
6  2017  Con Descuento  384397.1052   -10718.4426
7  2017  Sin Descuento  348818.1500   104157.7122

In [20]:
import plotly.express as px

In [21]:
import plotly.graph_objects as go

In [22]:
# Make sure to execute the cell containing this code first:
comparativa = df.groupby(['Year', 'flag_descuento']).agg(
    Total_Sales=('Sales', 'sum'),  # Sumar las ventas por año y descuento
    Total_Profit=('Profit', 'sum')  # Sumar los beneficios por año y descuento
).reset_index()

fig = go.Figure()

fig.add_trace(go.Bar(
    x=comparativa[comparativa['flag_descuento'] == 'Con Descuento']['Year'],
    y=comparativa[comparativa['flag_descuento'] == 'Con Descuento']['Total_Sales'],
    name='Ventas con Descuento',
    marker=dict(color='skyblue'),
))


fig.add_trace(go.Bar(
    x=comparativa[comparativa['flag_descuento'] == 'Sin Descuento']['Year'],
    y=comparativa[comparativa['flag_descuento'] == 'Sin Descuento']['Total_Sales'],
    name='Ventas sin Descuento',
    marker=dict(color='orange')
))


fig.add_trace(go.Scatter(
    x=comparativa[comparativa['flag_descuento'] == 'Con Descuento']['Year'],
    y=comparativa[comparativa['flag_descuento'] == 'Con Descuento']['Total_Profit'],
    name='Beneficio con Descuento',
    mode='lines+markers',
    line=dict(color='darkblue', width=2),
))


fig.add_trace(go.Scatter(
    x=comparativa[comparativa['flag_descuento'] == 'Sin Descuento']['Year'],
    y=comparativa[comparativa['flag_descuento'] == 'Sin Descuento']['Total_Profit'],
    name='Beneficio sin Descuento',
    mode='lines+markers',
    line=dict(color='red', width=2),
))


fig.update_layout(
    barmode='group',
    title='Ventas y Beneficio por Año y Tipo de Descuento',
    xaxis_title='Año',
    yaxis_title='Cantidad',
    xaxis={'type': 'category'},
    legend_title_text='Tipo de Descuento y Métrica',
    template='plotly_white'
)

fig.show()

En este gráfico comprobamos que el beneficio de ventas con descuento es mucho menor que los beneficios que se obtienen con las ventas sin descuento.
Esto resalta y nos hace pensar que no es tan beneficioso tener mayor ventas con descuento, ya que el beneficiio es infinitamenente menor.
Deberíasmos transmitir a la empresa que quizas deban de cambiar su filosofia de ventas ya que no ocasiona rentabilidad y a la larga puede perjudicar.
Esto genera mas trabajo logistico y menos beneficio.


Conclusiones: con esto podemos justificar a la empresa que si continua con la dinamica de aplicar los descuentos tendrá un menor benefico



<span style="font-size: 15pt; font-weight: bold; color: #0098cd;">2.3. Gráfico 3.</span>

Respuesta:

In [23]:
fig = px.histogram(
    df,
    x='Discount',
    title='Distribución de Descuentos',
    labels={'Discount': 'Descuento'}
)

fig.show()

Conclusiones: aquí le podemos mostrar el cliente la cantidad de productos que tiene con descuento y con que tipo de descuento aplicado

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">3. Guardado del Trabajo.</span>

Exportando una gráfica.

In [ ]:
#Guardado del Trabajo.
df.to_csv('Pontia_CasoPractico2.csv', index=False)
#Exportando una gráfica.
# En colab necesitais añadir en fig.show(renderer='notebook')
fig.show(renderer='notebook')


Transformando nuestro Notebook en un Reporting.

In [26]:
!jupyter nbconvert  --to html --no-input Pontia_CasoPractico2.ipynb

[NbConvertApp] WARNING | pattern 'Pontia_CasoPractico2.ipynb' matched no files
This application is used to convert notebook files (*.ipynb)
        to various other formats.


Options
The options below are convenience aliases to configurable class-options,
as listed in the "Equivalent to" description-line of the aliases.
To see all configurable class-options for some <cmd>, use:
    <cmd> --help-all

--debug
    set log level to logging.DEBUG (maximize logging output)
    Equivalent to: [--Application.log_level=10]
--show-config
    Show the application's configuration (human-readable format)
    Equivalent to: [--Application.show_config=True]
--show-config-json
    Show the application's configuration (json format)
    Equivalent to: [--Application.show_config_json=True]
--generate-config
    generate default config file
    Equivalent to: [--JupyterApp.generate_config=True]
-y
    Answer yes to any questions instead of prompting.
    Equivalent to: [--JupyterApp.answer_yes=True]
--ex

Nota: En colab necesitais añadir en fig.show(renderer='notebook')